proto/internal/temporal/server/api/taskqueue/v1/message.proto

```diff
@@ -80,16 +80,18 @@ message BuildIdRedirectInfo {
     string assigned_build_id = 1;
 }
 
-// Information about task forwarding from one partition to its parent. Versioning decisions are made
-// at the source partition and sent to the parent partition in this message.
+// Information about task forwarding from one partition to its parent. Versioning decisions for activity/workflow
+// tasks are made at the source partition and sent to the parent partition in this message so that parent partition
+// does not have to make versioning decision again. For Query/Nexus tasks, this works differently as the child's
+// versioning decision is ignored and the parent partition makes a fresh decision.
 message TaskForwardInfo {
     // RPC name of the partition forwarded the task.
     // In case of multiple hops, this is the source partition of the last hop.
     string source_partition = 1;
     temporal.server.api.enums.v1.TaskSource task_source = 2;
     // Redirect info is not present for Query and Nexus tasks.
     BuildIdRedirectInfo redirect_info = 3;
-    // Build ID that should be used to dispatch the task to.
+    // Build ID that should be used to dispatch the task to. Ignored in Query and Nexus tasks.
```

Comment:
I see that workflow/activity tasks don't have the same issue that query/nexus tasks do, but why can't we just always evaluate versioning in the parent? Then things are simpler and more consistent, there's less code and comments to understand. Are there any downsides?

Reply:
I don't see a downside other than it's a slightly bigger change. I just wanted to hold off on that until versioning designs are more clear.